test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java

```diff
@@ -43,42 +48,64 @@ enum XPackRequested {
 
         XPackRequested xpackRequested = XPackRequested.NOT_SPECIFIED;
 
-        public SkipSectionBuilder withVersion(String version) {
+        public SkipSectionBuilder skipVersion(String version) {
             this.version = version;
             return this;
         }
 
-        public SkipSectionBuilder withReason(String reason) {
+        public SkipSectionBuilder setReason(String reason) {
             this.reason = reason;
             return this;
         }
 
-        public SkipSectionBuilder withTestFeature(String featureName) {
-            this.testFeatures.add(featureName);
+        public SkipSectionBuilder withYamlRunnerFeature(String featureName) {
+            this.yamlRunnerFeatures.add(featureName);
+            return this;
+        }
+
+        public SkipSectionBuilder withXPack() {
+            if (xpackRequested == XPackRequested.NO) {
+                xpackRequested = XPackRequested.MISMATCHED;
+            } else {
+                xpackRequested = XPackRequested.YES;
+            }
             return this;
         }
 
-        public void withXPack(boolean xpackRequired) {
-            if (xpackRequired && xpackRequested == XPackRequested.NO || xpackRequired == false && xpackRequested == XPackRequested.YES) {
+        public SkipSectionBuilder skipXPack() {
+            if (xpackRequested == XPackRequested.YES) {
                 xpackRequested = XPackRequested.MISMATCHED;
             } else {
-                xpackRequested = xpackRequired ? XPackRequested.YES : XPackRequested.NO;
+                xpackRequested = XPackRequested.NO;
             }
+            return this;
+        }
+
+        public SkipSectionBuilder skipClusterFeature(String featureName) {
```

Comment:
This is effectively a double-skip. Even reading the code, I'm not sure what this and the method below actually means. Is there something else we could use, maybe using the `skip/lacks` terminology somehow?

Naming is hard :(

Reply:
Probably it's not a realistic option in terms of effort (and impact on people's routine). But we could consider refactoring the `skip` section into a `requires` (precondition like) section with positive checks. That would avoid the need for any double negation / double skipping. IMHO as long as the overall section remains negative, naming and understanding will always be troublesome.

Reply:
Actually, it looks like this isn't a double skip, but I think the above would still help 🤯 

Anyways, as alternative, would it be easier to read if the builder used `skipOnVersion` (`version`), `skipOnYamlRunnerFeatuerAbsent` (`feature`), `skipOnClusterFeature` (`cluster_has_features `) and `skipOnClusterFeatureAbsent` (`cluster_lacks_features`).

The problem remains that some fields (`version` vs `feature`) are using opposed semantics. That doesn't make it any easier to add support for cluster features.

Reply:
> Naming is hard :(

Oh I agree so much :) I think I have renamed these at least 2/3 times to try and make these understandable, and I'm still unsatisfied.
It does not help that the most common/useful scenario is expressed as a double negation: we are in a skip section, and we will want to skip a test if the cluster does _not_ support the feature we are testing. So, "we want the cluster to not have this feature to not run this test" skip if the cluster lacks something, to say "we want the cluster to have this feature to run this test".
It also does not help in the least that today we already have an anomaly in skip (`features`, as in test runner supported features - they are _positives_, i.e. they have to be present _otherwise_ the test will be skipped, in a section where everything else is negative).

I really tried to avoid it as it will involve a greater change, and I thought that it was better to keep things minimal and incremental, but maybe @mosche suggestion is the right way to go: we may really want to introduce a `requires` section, and introduce `cluster_features` in both (so it will be negative in one, positive in the other), and correct the aforementioned anomaly moving (test runner)`features` from `skip` to `requires` too. Keeping them in `skip` if we add `require` will just increase confusion.

Wdyt? Should I proceed with creating a separate `requires` section, or should we strive to keep just `skip` and find better names?

Reply:
I like @mosche suggestion of `skipOnFeaturePresence` and `skipOnFeatureAbsence` - that makes it a bit clearer, I think.

Reply:
I think that would be really beneficial and avoid any further confusion! 💯

Reply:
> Actually, it looks like this isn't a double skip

Correct: I tried to use `with` for positive (skip if `not`, e.g. `withXPack` -> we want XPack -> skip if _not_ present) and `skip` for negative (`skipVersion` -> skip if version is in the range specified).

I like your proposal to express all in terms of skip and explicit present/absent - less brain jumps. So `withXPack` -> `skipIfXPackAbsent` and `skipXPack` -> `skipIfXPackPresent`)

We can have `skipIfClusterFeaturesPresent` and `skipIfClusterFeaturesAbsent`, and just `skipIfYamlRunnerFeatuerAbsent` (to reflect what is present today).

Reply:
@thecoop How would you name the fields in Yaml then? According to the builder methods?
The double twist remains obviously, but it feels slightly less confusing.

```yaml
skip:
  - version:
  # rename to test_feature_absence?
  - feature:
  - cluster_feature_presence:
  - cluster_feature_absence:
```

Reply:
Let me do it this way: I'll push a new revision to this PR with the renaming. 
Then, I'll post a draft with `requires`, to see how large a change is and how we like it. I'll spend a limited time on it (e.g. tomorrow); if it's more, we need to reconsider (but we'll know it's more and what does it require)

Reply:
Instead of the skip terminology, I wonder if we could use positive terminology of "require", eg "requireXpack". Alternatively, we could use the same terminology as Junit, using "assumeXXX", which implies a skip of the assumption is violated.

Reply:
I've renamed them as @mosche suggested, but if we are all aligned and prefer `required` for positive and `skip` for negative (e.g. `requireXPack`, `skipXPack`) I can rename them again.
I kind of like the current naming more though - after all, this is `SkipSection`. I'd definitely go with `requireXxx` if we want to add a `RequireSection`.